### DISCUSSION TITLE
Replacement for Postgres SERIAL (deprecation of IDENTITY)

### DISCUSSION THREAD

**Discussion by aleho** [doctrine/doctrine2#10781] [Support Questions] [answered]
I'm getting a deprecation notice (#9833) when using `#[GeneratedValue(strategy: 'IDENTITY')]`, which according to docs results in the non-portable `SERIAL`, so I tried to follow the recommended changes.

Using `strategy: 'AUTO'` (with the suggested DB changes) I've now come across undocumented behavior that breaks quite badly here.

Apparently as soon as an entity is persisted it will a) be assigned an ID and b) considered "flushed" even if there wasn't a single flush in the entire process of `$e = new Entity(); $em->persist($e);`. This affects code that either checks for IDs (bad) or calls `EntityManager::contains($e)` (good?) to know whether an entity is new, because now it gets the wrong answers on Postgres.

On MariaDB/MySQL `EntityManager::contains($e)` returns false unless the entity has been flushed.

- Is there a reliable way to find out whether an entity has been "written" to the DB on Postgres?
- Is `EntityManager::contains()` even supposed to work?
- How should we update code that relies on finding out if an entity that has been persisted is actually in the DB? Call `find` on the repository? Keep a local pool of newly created entities and empty it on `flush`?

The deprecation notice I get is:

```
User Deprecated: Context: Loading metadata for class App\Entity\Config\Feed Problem: Using the IDENTITY generator strategy with platform "Doctrine\DBAL\Platforms\PostgreSQL100Platform" is deprecated and will not be possible in Doctrine ORM 3.0. Solution: Use the SEQUENCE generator strategy instead. (ClassMetadataFactory.php:627 called by ClassMetadataFactory.php:147, https://github.com/doctrine/orm/issues/8850, package doctrine/orm)
```

This is somewhat related to #10739 but I didn't want to hijack an already answered question.

**Comment by greg0ire** [comment]


**Reply by aleho** [reply]
The deprecation notice I get is:

```
User Deprecated: Context: Loading metadata for class App\Entity\Config\Feed Problem: Using the IDENTITY generator strategy with platform "Doctrine\DBAL\Platforms\PostgreSQL100Platform" is deprecated and will not be possible in Doctrine ORM 3.0. Solution: Use the SEQUENCE generator strategy instead. (ClassMetadataFactory.php:627 called by ClassMetadataFactory.php:147, https://github.com/doctrine/orm/issues/8850, package doctrine/orm)
```

This lead me to believe that using `SEQUENCE` (or `AUTO` resolving to `SEQUENCE`) on Postgres was the recommended action.

Was I wrong? Is this message related to internal changes and not meant to be actionable by users?


There was also

```
User Deprecated: Doctrine\DBAL\Platforms\PostgreSQLPlatform::usesSequenceEmulatedIdentityColumns is deprecated. (PostgreSQLPlatform.php:213 called by ClassMetadataFactory.php:626, https://github.com/doctrine/dbal/pull/5513, package doctrine/dbal)
```

which I, afaik, didn't investigate further because I was focusing on the notice above.

**Reply by greg0ire** [reply]
> Was I wrong? Is this message related to internal changes and not meant to be actionable by users?

Nope I was wrong, sorry, it does look like you should avoid `IDENTITY` here :thinking:

**Reply by aleho** [reply]
So what's the recommended way now to find out if a not yet flushed entity is "new" even though it has an ID and `EntityManager::contains()` returns `true`?

**Reply by greg0ire** [reply]
I don't know, sorry, I will delete my answer for now.

**Comment by greg0ire** [comment, marked as answer]
> Is there a reliable way to find out whether an entity has been "written" to the DB on Postgres?

I'm assuming you mean this only for new entities.
I might be wrong, but isn't that what `$entityManager->getUnitOfWork()->isScheduleForInsert($entity)` would do?

> Is EntityManager::contains() even supposed to work?
> On MariaDB/MySQL EntityManager::contains($e) returns false unless the entity has been flushed.

I find this kind of hard to believe. Can you write a test case reproducing the issue?

**Reply by aleho** [reply]
Sorry for the late reply. I made a reproducer, but wasn't able to actually reproduce it in MariaDB. Don't know why or what caused it. Most likely an error on my side.

So, `isScheduledForInsert` is the recommended way to check if an entity is new without relying on the generated ID of that entity (after persist, before flush)?

**Reply by greg0ire** [reply]
I don't know that there is an "officially recommended" :tm: way of doing this, but I'm recommended it to you. That recommendation is provided "as is", without any warranty of any kind etc. etc.

**Reply by aleho** [reply]
Thanks, I'll take that! My tests in the reproducer were OK across MariaDB an Postgres, which is all I'm asking for.
